File: app/export/hpgl_exporter.py
"""HPGL export (DrawingBot V3 plotter format)."""

from __future__ import annotations

from typing import Dict, List

from app.core.drawing_area import DrawingArea
from app.core.drawing_pen import DrawingPen
from app.export.path_export import prepare_pen_paths


def _mm_to_hpgl(x_mm: float, y_mm: float, da: DrawingArea, hpgl: Dict) -> tuple[int, int]:
    """Map mm coordinates to HPGL plotter units."""
    w_mm = max(da.width_mm, 0.001)
    h_mm = max(da.height_mm, 0.001)
    x_max = int(hpgl.get("x_max", 16158))
    y_max = int(hpgl.get("y_max", 11040))

    x = int(x_mm / w_mm * x_max)
    y = int(y_mm / h_mm * y_max)

    if hpgl.get("x_mirror", False):
        x = x_max - x
    if hpgl.get("y_mirror", False):
        y = y_max - y

    return x, y


def export_hpgl(
    filepath: str,
    geometries,
    drawing_area: DrawingArea,
    active_pens: List[DrawingPen],
    hpgl_settings: Dict,
    path_opt: Dict,
) -> None:
    pen_paths = prepare_pen_paths(geometries, path_opt)
    h_mm = drawing_area.height_mm
    pen_num = int(hpgl_settings.get("pen_number", 1))
    velocity = int(hpgl_settings.get("pen_velocity", 10))
    flatness = float(hpgl_settings.get("curve_flatness", 0.1))

    lines = [
        "IN;",
        f"SP{pen_num};",
        f"VS{velocity};",
        f"FS{flatness};",
    ]

    for pi in sorted(pen_paths.keys()):
        pen = active_pens[pi % len(active_pens)]
        lines.append(f"; Layer: {pen.name}")
        for path in pen_paths[pi]:
            if len(path) < 2:
                continue
            x0, y0 = path[0]
            hx, hy = _mm_to_hpgl(x0, h_mm - y0, drawing_area, hpgl_settings)
            lines.append(f"PU{hx},{hy};")
            segments = []
            for x, y in path[1:]:
                hx, hy = _mm_to_hpgl(x, h_mm - y, drawing_area, hpgl_settings)
                segments.append(f"{hx},{hy}")
            if segments:
                lines.append(f"PD{','.join(segments)};")

    lines.append("SP0;")
    with open(filepath, "w", encoding="utf-8") as f:
        f.write("\n".join(lines))

File: cpp_backup/export/hpgl_exporter.cpp
#include "export/hpgl_exporter.h"
#include "export/path_optimizer.h"
#include <QFile>
#include <QTextStream>
#include <QStringList>

static std::pair<int, int> mmToHPGL(double xMm, double yMm, const DrawingAreaConfig& da, const HPGLSettings& settings) {
    double wMm = std::max(da.width_mm, 0.001);
    double hMm = std::max(da.height_mm, 0.001);

    int x = static_cast<int>((xMm / wMm) * settings.xMax);
    int y = static_cast<int>((yMm / hMm) * settings.yMax);

    if (settings.xMirror) x = settings.xMax - x;
    if (settings.yMirror) y = settings.yMax - y;

    return {x, y};
}

bool HPGLExporter::exportHPGL(const QString& filepath,
                              const QVector<DrawingGeometry>& geometries,
                              const DrawingAreaConfig& drawingArea,
                              int imageWidth,
                              int imageHeight,
                              const HPGLSettings& settings) {
    if (geometries.isEmpty()) return false;

    QFile file(filepath);
    if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        return false;
    }

    QTextStream out(&file);

    // Convert pixel coordinates to physical mm
    QVector<DrawingGeometry> mmGeoms;
    mmGeoms.reserve(geometries.size());
    for (const auto& geom : geometries) {
        Path mmPath = pixel_path_to_mm(geom.path, imageWidth, imageHeight, drawingArea);
        if (!mmPath.empty()) {
            mmGeoms.append({mmPath, geom.penIndex, geom.groupIndex});
        }
    }

    // Prepare, filter and optimize paths
    QMap<int, QVector<Path>> penPaths = PathOptimizer::preparePenPaths(
        mmGeoms, settings.minPathLength, settings.optimizePaths);

    double h_mm = drawingArea.height_mm;

    out << "IN;\n";
    out << QString("SP%1;\n").arg(settings.penNumber);
    out << QString("VS%1;\n").arg(settings.penVelocity);
    out << QString("FS%1;\n").arg(settings.curveFlatness, 0, 'f', 2);

    for (auto it = penPaths.begin(); it != penPaths.end(); ++it) {
        int penIdx = it.key();
        out << "; Layer: Pen " << penIdx << "\n";

        for (const auto& path : it.value()) {
            if (path.size() < 2) continue;

            auto [hx0, hy0] = mmToHPGL(path[0].first, h_mm - path[0].second, drawingArea, settings);
            out << QString("PU%1,%2;\n").arg(hx0).arg(hy0);

            QStringList segments;
            for (std::size_t i = 1; i < path.size(); ++i) {
                auto [hx, hy] = mmToHPGL(path[i].first, h_mm - path[i].second, drawingArea, settings);
                segments.append(QString("%1,%2").arg(hx).arg(hy));
            }

            if (!segments.isEmpty()) {
                out << "PD" << segments.join(",") << ";\n";
            }
        }
    }

    out << "SP0;\n";
    return true;
}
